### ISSUE TITLE
`PhantomPinned` warns that field is never read, but `PhantomData` does not

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/rust#154888]
I tried this code:

```rust
use core::marker::{PhantomData, PhantomPinned};

struct Test {
    pin: PhantomPinned,
    data: PhantomData<()>,
}

fn main() {
    let _ = Test { pin: PhantomPinned, data: PhantomData };
}
```

This should not show any errors indicating that `pin` was not read, since its purpose is merely to avoid an `Unpin` impl. Instead, I get this output:

```
warning: field `pin` is never read
 --> src/main.rs:4:5
  |
3 | struct Test {
  |        ---- field in this struct
4 |     pin: PhantomPinned,
  |     ^^^
  |
  = note: `#[warn(dead_code)]` (part of `#[warn(unused)]`) on by default
```

Note that `PhantomData` is specially handled to not do this, and I would expect `PhantomPinned` to act the same.

### Meta

`rustc --version --verbose`:
```
rustc 1.96.0-nightly (55e86c996 2026-04-02)
binary: rustc
commit-hash: 55e86c996809902e8bbad512cfb4d2c18be446d9
commit-date: 2026-04-02
host: x86_64-unknown-linux-gnu
release: 1.96.0-nightly
LLVM version: 22.1.2
```

**Comment by asder8215** [comment]
@rustbot claim

This seems to be a clippy problem. I notice in [`clippy_lints/src/missing_fields_in_debug.rs`](https://github.com/rust-lang/rust/blob/a92a99ef1479a9987b0ef76a064c2bd8e779babd/src/tools/clippy/clippy_lints/src/missing_fields_in_debug.rs#L187) it avoids emitting a field is not used warning specifically for `PhantomData`. We'd also have to modify [`clippy_lints/src/pub_underscore_fields.rs`](https://github.com/rust-lang/rust/blob/a92a99ef1479a9987b0ef76a064c2bd8e779babd/src/tools/clippy/clippy_lints/src/pub_underscore_fields.rs#L79) since it also excludes `PhantomData` from emitting a warning if a field is declared as public but uses a leading underscore.

In both cases, clippy does have access to checking if the marker is `PhantomData` through `LangItem::PhantomData`, but from the [`language_item_table`](https://github.com/rust-lang/rust/blob/a92a99ef1479a9987b0ef76a064c2bd8e779babd/compiler/rustc_hir/src/lang_items.rs#L153), it doesn't seem like there's a lang item for `PhantomPinned` and it's not marked with that `#[lang = "..."]` macro in [here](https://github.com/rust-lang/rust/blob/a92a99ef1479a9987b0ef76a064c2bd8e779babd/library/core/src/marker.rs#L1024) just like how [`PhantomData`](https://github.com/rust-lang/rust/blob/a92a99ef1479a9987b0ef76a064c2bd8e779babd/library/core/src/marker.rs#L809) is.

Is it okay to add `PhantomPinned` as a `lang_item` in the `language_item_table`?

**Comment by asder8215** [comment]
> For now, the attribute could be something like `#[rustc_marker_type]`, and it would tag the field as not applying for dead code checks. But maybe in the future, we could have it be some sort of diagnostic attribute if we find a use case for it.

Just so I'm understanding you correctly, are you suggesting that I create a built in attribute called `rustc_marker_type` within [`compiler/rustc_feature/src/builtin_attrs.rs`](https://github.com/rust-lang/rust/blob/main/compiler/rustc_feature/src/builtin_attrs.rs)? I've never written code on the compiler side of Rust, so I'm just going off intuition on where everything is at and what my go to references are telling me in my IDE.

The way I'm thinking about how your approach would work is that we just add the `rustc_marker_type` in `compiler/rustc_feature/src/builtin_attrs.rs`, create the `rustc_marker_type` symbol inside `compiler/rustc_span/src/symbol.rs`, and just add that macro on top of `PhantomPinned` (and any of the other `Phantom*` if we want to do that). The only thing that I'm trying to figure out is how the two clippy lint files I mentioned earlier can access information about what attributes are implemented on a struct. I know `is_lang_item()` is used in the file to check if the type of the field is a specific `LangItem`. Does `is_diag_item()` do the trick here?

> Just my 2¢, but it might be worth having this operate via an attribute instead of adding an extra lang item for `PhantomPinned`, since it makes sense that potentially other types could be added that want similar handling. For example, there are a bunch of `Phantom*` types that internally wrap a `PhantomData` and I'm not sure if those count for the diagnostics either.

And it makes sense to me to use an attribute over adding an extra lang item. We'd have less symbols to add using an attribute as well.

On that note, I've been wondering if [`rustc_trivial_field_reads`](https://github.com/rust-lang/rust/blob/bcded331651b60a0383b3ff51db4f24c4495ac53/compiler/rustc_feature/src/builtin_attrs.rs#L1158) is related to the attribute you were proposing? I notice it mentions in the comments above:
```rust
// Prevents field reads in the marked trait or method to be considered
// during dead code analysis.
```
But I'm unsure if this is unrelated.
